---
title: "**ChIP-seq analyses**"
subtitle: Analyses and figures
author: "Marta Coronado Zamora"
date: "4 September 2023"
output:
  html_document:
    theme: yeti
    css: https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/all.css
    self_contained: yesR
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, fig.width = 6, fig.height = 4, fig.align = "center")
```

# Libraries

```{r requiredLibraries, message=FALSE}

library(ggplot2)
library(ggpubr)
library(VennDiagram)
library(wesanderson)
library(data.table)
library(dplyr)
library(tidyr)
library(cowplot)
library(ggvenn)
library(ggthemes)
library(distributions3)
library(UpSetR)
```

# Functions

```{r functions, message=FALSE}

source("vennDiagram.R")
source("ggplot_theme.R")
```

<div style="background-color: #86CBBB; 1px; height:3px " ></div><br>

# Epigenetic states of TEs on flanking sequences is body part-dependent

Analysis with all TEs (4,823 TEs).

```{r analysis all TEs - figure 1}
TEtrack <- fread("intersectTEsWindowsSignalValueTrack.sort.bed")
tmpTE <- TEtrack  %>%  dplyr::select(V1,V4,V8,V9,V14,V15,V16,V17)
colnames(tmpTE) <- c("chr","TE", "position", "type","signalValue","tissue","strain","histone")
dfPlot <- tmpTE
dfPlot <- dfPlot[dfPlot$position!="TE",]
dfPlot$position <- factor(dfPlot$position, levels=c("-20kb", "-19kb", "-18kb", "-17kb", "-16kb", "-15kb", "-14kb", "-13kb", "-12kb", "-11kb", "-10kb", "-9kb", "-8kb", "-7kb", "-6kb", "-5kb", "-4kb", "-3kb", "-2kb", "-1kb", "1kb", "2kb", "3kb", "4kb", "5kb", "6kb", "7kb", "8kb", "9kb", "10kb", "11kb", "12kb", "13kb", "14kb", "15kb", "16kb", "17kb", "18kb", "19kb", "20kb"))
dfPlotMedian <- dfPlot %>% group_by(position,type,tissue,histone) %>% summarise(meanSV = median(signalValue))
dfPlotMedian <- dfPlot %>% group_by(position,type,tissue,strain,histone) %>% summarise(meanSV = median(signalValue))
dfPlotMedian$windows <- c(rev(rep(paste0("-",seq(1,20,1)),each=6)),rep(seq(1,20,1),each=6))
dfPlotMedian$windows <- c(rev(rep(paste0("-",seq(1,20,1)),each=30)),rep(seq(1,20,1),each=30))
dfPlotMedian$windows <- as.integer(dfPlotMedian$windows)

dfPlotMedian <- fread("epigenetic_landscapes/dfPlotMedian.allTE.tsv")
dfPlotMedian[dfPlotMedian$tissue=="head",]$tissue<-"Head"
dfPlotMedian[dfPlotMedian$tissue=="gut",]$tissue<-"Gut"
dfPlotMedian[dfPlotMedian$tissue=="ovary",]$tissue<-"Ovary"
dfPlotMedian$tissue <- factor(dfPlotMedian$tissue,levels = c("Head","Gut","Ovary"))
dfPlotMedian$histone <- factor(dfPlotMedian$histone,levels = c("H3K9me3","H3K27ac"))

figure1_a <- ggplot(dfPlotMedian, aes(x=windows,y=meanSV)) + geom_vline(aes(xintercept=0), color="grey30")  + geom_smooth(method="loess", span = 0.1, aes(color=tissue))  +
theme_light(base_size = 18) + scale_color_tableau() + labs(x="Distance to the TE (kb)", y = paste0("Median histone fold enrichment"), color="Body part")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10)) + scale_x_continuous(breaks = c(-20,-15,-10,-5,0,5,10,15,20), labels=c("-20","-15","-10","-5","TE","5","10","15","20"), minor_breaks = seq(-20, 20, 1)) +  
theme(panel.grid.minor.x = element_line(colour="grey90", linewidth=0.5, linetype = "dashed"), panel.grid.major.x = element_line(colour="grey90", linewidth = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + facet_grid(.~histone) +
  theme(legend.position = "bottom")
figure1_a
ggsave("FIGURES/Figure_1A.pdf", plot = figure1_a, width = 9, height = 6)
save(file="FIGURES/Figure_1A.RData", figure1_a, dfPlotMedian)

dfPlotMedian <- fread("epigenetic_landscapes/dfPlotMedian.allTE.strain.tsv")
dfPlotMedian[dfPlotMedian$tissue=="head",]$tissue<-"Head"
dfPlotMedian[dfPlotMedian$tissue=="gut",]$tissue<-"Gut"
dfPlotMedian[dfPlotMedian$tissue=="ovary",]$tissue<-"Ovary"
dfPlotMedian$tissue <- factor(dfPlotMedian$tissue,levels = c("Head","Gut","Ovary"))
dfPlotMedian$histone <- factor(dfPlotMedian$histone,levels = c("H3K9me3","H3K27ac"))

figure_s1 <- ggplot(dfPlotMedian, aes(x=windows,y=meanSV)) + geom_vline(aes(xintercept=0), color="grey30")  + geom_smooth(method="loess", span = 0.1, aes(color=tissue))  +
theme_light(base_size = 18) + scale_color_tableau() + labs(x="Distance to the TE (kb)", y = paste0("Median histone fold enrichment"), color="Body part")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10)) + scale_x_continuous(breaks = c(-20,-15,-10,-5,0,5,10,15,20), labels=c("-20","-15","-10","-5","TE","5","10","15","20"), minor_breaks = seq(-20, 20, 1)) +  
theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + facet_grid(strain~histone)
figure_s1
ggsave("FIGURES/Figure_S1.pdf", plot = figure_s1, width = 10, height = 10)
save(file="FIGURES/Figure_S1.RData", figure_s1, dfPlotMedian)
```

Analysis with the TEs that are polymorphic and non-nested: 2,325.

```{r polymorphic-TEs}
TEtrack <- fread("intersectTEsFilterBreakpointsWindowsSignalValueTrack.sort.clean.bed")
breakpointsTrack <- fread("intersectBreakpointsWindowsSignalValueTrack.sort.clean.bed")

tmp<-breakpointsTrack  %>%  dplyr::select(V1,V4,V7,V8,V13,V14,V15,V16)
colnames(tmp) <- c("chr","TE", "position", "type","signalValue","tissue","strain","histone")

tmpTE<-TEtrack  %>%  dplyr::select(V1,V4,V8,V9,V14,V15,V16,V17)
colnames(tmpTE) <- c("chr","TE", "position", "type","signalValue","tissue","strain","histone")
  
dfPlot <- rbind(tmpTE, tmp)

dfPlot <- dfPlot[dfPlot$position!="breakpoint",]
dfPlot <- dfPlot[dfPlot$position!="TE",]
  
dfPlot$position <- factor(dfPlot$position, levels=c("-20kb", "-19kb", "-18kb", "-17kb", "-16kb", "-15kb", "-14kb", "-13kb", "-12kb", "-11kb", "-10kb", "-9kb", "-8kb", "-7kb", "-6kb", "-5kb", "-4kb", "-3kb", "-2kb", "-1kb", "1kb", "2kb", "3kb", "4kb", "5kb", "6kb", "7kb", "8kb", "9kb", "10kb", "11kb", "12kb", "13kb", "14kb", "15kb", "16kb", "17kb", "18kb", "19kb", "20kb"))
write.table(dfPlot,file="dfPlot.allTE-breakpoint.clean.tsv",col.names = T, sep = "\t", row.names = F, quote = F)
dfPlot <- fread("dfPlot.allTE-breakpoint.clean.tsv")

# enrichment
# gut, H3K9me3
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K9me3", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)
# ovary, H3K9me3
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K9me3", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)
# head, H3K9me3
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K9me3", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K9me3", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)

# gut, H3K27ac
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K27ac", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "gut", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)
# ovary, H3K27ac
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K27ac", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "ovary", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)
# head, H3K27ac
round(( dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K27ac", type=="TE")  %>% pull(signalValue) %>% median() - dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() ) /  dfPlot  %>% filter(position == "-1kb" | position == "1kb") %>% filter(tissue == "head", histone == "H3K27ac", type=="breakpoint")  %>% pull(signalValue) %>% median() * 100,3)
#

dfPlotMedian <- dfPlot %>% group_by(position,type,tissue,histone) %>% summarise(meanSV = median(signalValue))
dfPlotMedianStrain <- dfPlot %>% group_by(position,type,tissue,histone,strain) %>% summarise(meanSV = median(signalValue))
#myLoc <- (which(levels(dfPlot$V8) == "-1kb") + which(levels(dfPlot$V8) == "1kb")) / 2

dfPlotMedian$windows <- c(rev(rep(paste0("-",seq(1,20,1)),each=12)),rep(seq(1,20,1),each=12))
dfPlotMedianStrain$windows <- c(rev(rep(paste0("-",seq(1,20,1)),each=60)),rep(seq(1,20,1),each=60))

dfPlotMedian$windows <- as.integer(dfPlotMedian$windows)
dfPlotMedianStrain$windows <- as.integer(dfPlotMedianStrain$windows)

write.table(dfPlotMedian,"dfPlotMedian.allTE-breakpoint.clean.tsv", quote=F, row.names = F, col.names = T)
write.table(dfPlotMedianStrain,"dfPlotMedian.allTE-breakpoint.strain.clean.tsv", quote=F, row.names = F, col.names = T)

# Figure 2
dfPlotMedian <- fread("epigenetic_landscapes/dfPlotMedian.allTE-breakpoint.clean.tsv")

dfPlotMedian[dfPlotMedian$tissue=="head",]$tissue<-"Head"
dfPlotMedian[dfPlotMedian$tissue=="gut",]$tissue<-"Gut"
dfPlotMedian[dfPlotMedian$tissue=="ovary",]$tissue<-"Ovary"
dfPlotMedian$tissue <- factor(dfPlotMedian$tissue,levels = c("Head","Gut","Ovary"))
dfPlotMedian$histone <- factor(dfPlotMedian$histone,levels = c("H3K9me3","H3K27ac"))

# All
figure_1b <- ggplot(dfPlotMedian, aes(x=windows,y=meanSV)) + geom_vline(aes(xintercept=0), color="grey30")  + geom_smooth(method="loess", span = 0.1, aes(color=type, group=type))  +
  theme_light(base_size = 18) + scale_color_manual(values=c("mistyrose4","darkred"),breaks = c("breakpoint","TE"),labels=c("TE(-)","TE(+)")) + labs(x="Distance to the TE (kb)", y = paste0("Median histone fold enrichment"), color="Type")   + scale_y_continuous(breaks = scales::pretty_breaks(n = 6)) + scale_x_continuous(breaks = c(-20,-15,-10,-5,0,5,10,15,20), labels=c("-20","-15","-10","-5","TE","5","10","15","20"), minor_breaks = seq(-20, 20, 1)) +  scale_linetype(guide = 'none') + 
  theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + facet_grid(tissue~histone, scales = "free_y") + theme(legend.position = "bottom")
figure_1b
ggsave("FIGURES/Figure_1B.pdf", plot = figure_1b, width = 9, height = 6)
save(file="FIGURES/Figure_1B.RData", figure_1b, dfPlotMedian)


dfPlotMedian <- fread("epigenetic_landscapes/dfPlotMedian.allTE-breakpoint.strain.clean.tsv")
dfPlotMedian[dfPlotMedian$tissue=="head",]$tissue<-"Head"
dfPlotMedian[dfPlotMedian$tissue=="gut",]$tissue<-"Gut"
dfPlotMedian[dfPlotMedian$tissue=="ovary",]$tissue<-"Ovary"
dfPlotMedian$tissue <- factor(dfPlotMedian$tissue,levels = c("Head","Gut","Ovary"))
dfPlotMedian$histone <- factor(dfPlotMedian$histone,levels = c("H3K9me3","H3K27ac"))
figure_s2_a <- ggplot(dfPlotMedian[dfPlotMedian$histone=="H3K27ac"], aes(x=windows,y=meanSV)) + geom_vline(aes(xintercept=0), color="grey30")  + geom_smooth(method="loess", span = 0.1, aes(color=type, group=type))  + theme_light(base_size = 16)  + scale_color_manual(values=c("mistyrose4", "darkred"),breaks = c("breakpoint","TE"),labels=c("TE(-)","TE(+)"))  + labs(x="Distance to the TE (kb)", y = paste0("Median H3K27ac fold enrichment"), color="Type")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 4)) + scale_x_continuous(breaks = c(-20,-15,-10,-5,0,5,10,15,20), labels=c("-20","-15","-10","-5","TE","5","10","15","20"), minor_breaks = seq(-20, 20, 1)) +  theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + facet_grid(strain~tissue)
figure_s2_a
figure_s2_b <- ggplot(dfPlotMedian[dfPlotMedian$histone=="H3K9me3"], aes(x=windows,y=meanSV)) + geom_vline(aes(xintercept=0), color="grey30")  + geom_smooth(method="loess", span = 0.1, aes(color=type, group=type))  + theme_light(base_size = 16)  + scale_color_manual(values=c("mistyrose4", "darkred"),breaks = c("breakpoint","TE"),labels=c("TE(-)","TE(+)"))  + labs(x="", y = paste0("Median H3K9me3 fold enrichment"), color="Type")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 4)) + scale_x_continuous(breaks = c(-20,-15,-10,-5,0,5,10,15,20), labels=c("-20","-15","-10","-5","TE","5","10","15","20"), minor_breaks = seq(-20, 20, 1)) +  theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + facet_grid(strain~tissue)
figure_s2_b
ggsave("FIGURES/Figure_S2.pdf", plot = ggarrange(figure_s2_b,figure_s2_a, common.legend = T, legend = "bottom", nrow = 2, labels = "AUTO", font.label = list(size = 18)), width = 9, height = 12)
save(file="FIGURES/Figure_S2.RData", figure_s2_b, figure_s2_a, dfPlotMedian)
```

```{r}
save(file="FIGURES/Figure_1.RData", figure1_a, figure_1b)
ggsave("FIGURES/Figure_1.pdf", plot = ggarrange(figure1_a,figure_1b, widths = c(1,1.1), labels = "AUTO", font.label = list(size = 18)), width = 14, height = 7.5)
```

Analysis of TEs with epigenetic effects.
```{r}
epigeneticEffectsTes <- fread("epigenetic_effects/epigeneticEffectsTesParsed_v5.tab")

TEs_epigenetic_effects <- unique(epigeneticEffectsTes[epigeneticEffectsTes$effect=="increment" | epigeneticEffectsTes$effect=="depletion",]$TE)

length(unique(epigeneticEffectsTes[!epigeneticEffectsTes$TE %in% TEs_epigenetic_effects & epigeneticEffectsTes$effect!="mix",]$TE))

length(unique(epigeneticEffectsTes[!epigeneticEffectsTes$TE %in% TEs_epigenetic_effects & epigeneticEffectsTes$effect=="mix",]$TE))

length(unique(epigeneticEffectsTes[epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3",]$TE))

head_increment_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "increment",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

head_depletion_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "depletion",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

head_increment_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "increment",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

head_depletion_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "depletion",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

head_mix_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "mix") %>%
  reframe(unique(TE)) %>% pull()

head_increment_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "increment",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

head_depletion_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "depletion",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

head_no_enrichment_TE_list <- epigeneticEffectsTes %>% filter(tissue == "head",
                                effect == "noEnrichment",
                                !TE %in% head_depletion_H3K27ac_TE_list,
                                !TE %in% head_increment_H3K27ac_TE_list,
                                !TE %in% head_depletion_H3K9me3_TE_list,
                                !TE %in% head_increment_H3K9me3_TE_list,
                                !TE %in% head_mix_TE_list) %>%
  reframe(unique(TE)) %>% pull()

upset(fromList(list(head_increment_H3K9me3_TE_list=head_increment_H3K9me3_TE_list,head_depletion_H3K9me3_TE_list=head_depletion_H3K9me3_TE_list,head_increment_H3K27ac_TE_list=head_increment_H3K27ac_TE_list,head_depletion_H3K27ac_TE_list=head_depletion_H3K27ac_TE_list,head_mix_TE_list=head_mix_TE_list,head_increment_bivalent_TE_list=head_increment_bivalent_TE_list,head_depletion_bivalent_TE_list=head_depletion_bivalent_TE_list,head_no_enrichment_TE_list=head_no_enrichment_TE_list)), nsets = 20)

# Create data frames for each condition
head_H3K9me3_increment <- data.frame(tissue = "Head", type = "H3K9me3 increment", n = length(head_increment_H3K9me3_TE_list))
head_H3K9me3_depletion <- data.frame(tissue = "Head", type = "H3K9me3 depletion", n = length(head_depletion_H3K9me3_TE_list))
head_H3K27ac_increment <- data.frame(tissue = "Head", type = "H3K27ac increment", n = length(head_increment_H3K27ac_TE_list))
head_H3K27ac_depletion <- data.frame(tissue = "Head", type = "H3K27ac depletion", n = length(head_depletion_H3K27ac_TE_list))
head_Mix <- data.frame(tissue = "Head", type = "Mix", n = length(head_mix_TE_list))
head_Bivalent_increment <- data.frame(tissue = "Head", type = "Bivalent increment", n = length(head_increment_bivalent_TE_list))
head_Bivalent_depletion <- data.frame(tissue = "Head", type = "Bivalent depletion", n = length(head_depletion_bivalent_TE_list))
head_No_enrichment <- data.frame(tissue = "Head", type = "No enrichment", n = length(head_no_enrichment_TE_list))

# Combine the data frames into one
head <- bind_rows(
  head_H3K9me3_increment,
  head_H3K9me3_depletion,
  head_H3K27ac_increment,
  head_H3K27ac_depletion,
  head_Mix,
  head_Bivalent_increment,
  head_Bivalent_depletion,
  head_No_enrichment
)

head_No_consistent <- data.frame(tissue = "Head", type = "No consistent", n = 2325-sum(head$n))

head <- bind_rows(head, head_No_consistent)

gut_increment_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "increment",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

gut_depletion_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "depletion",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

gut_increment_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "increment",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

gut_depletion_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "depletion",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

gut_mix_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "mix") %>%
  reframe(unique(TE)) %>% pull()

gut_increment_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "increment",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

gut_depletion_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "depletion",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

gut_no_enrichment_TE_list <- epigeneticEffectsTes %>% filter(tissue == "gut",
                                effect == "noEnrichment",
                                !TE %in% gut_depletion_H3K27ac_TE_list,
                                !TE %in% gut_increment_H3K27ac_TE_list,
                                !TE %in% gut_depletion_H3K9me3_TE_list,
                                !TE %in% gut_increment_H3K9me3_TE_list,
                                !TE %in% gut_mix_TE_list) %>%
  reframe(unique(TE)) %>% pull()

upset(fromList(list(gut_increment_H3K9me3_TE_list=gut_increment_H3K9me3_TE_list,gut_depletion_H3K9me3_TE_list=gut_depletion_H3K9me3_TE_list,gut_increment_H3K27ac_TE_list=gut_increment_H3K27ac_TE_list,gut_depletion_H3K27ac_TE_list=gut_depletion_H3K27ac_TE_list,gut_mix_TE_list=gut_mix_TE_list,gut_increment_bivalent_TE_list=gut_increment_bivalent_TE_list,gut_depletion_bivalent_TE_list=gut_depletion_bivalent_TE_list,gut_no_enrichment_TE_list=gut_no_enrichment_TE_list)), nsets = 20)

# Create data frames for each condition
gut_H3K9me3_increment <- data.frame(tissue = "Gut", type = "H3K9me3 increment", n = length(gut_increment_H3K9me3_TE_list))
gut_H3K9me3_depletion <- data.frame(tissue = "Gut", type = "H3K9me3 depletion", n = length(gut_depletion_H3K9me3_TE_list))
gut_H3K27ac_increment <- data.frame(tissue = "Gut", type = "H3K27ac increment", n = length(gut_increment_H3K27ac_TE_list))
gut_H3K27ac_depletion <- data.frame(tissue = "Gut", type = "H3K27ac depletion", n = length(gut_depletion_H3K27ac_TE_list))
gut_Mix <- data.frame(tissue = "Gut", type = "Mix", n = length(gut_mix_TE_list))
gut_Bivalent_increment <- data.frame(tissue = "Gut", type = "Bivalent increment", n = length(gut_increment_bivalent_TE_list))
gut_Bivalent_depletion <- data.frame(tissue = "Gut", type = "Bivalent depletion", n = length(gut_depletion_bivalent_TE_list))
gut_No_enrichment <- data.frame(tissue = "Gut", type = "No enrichment", n = length(gut_no_enrichment_TE_list))

# Combine the data frames into one
gut <- bind_rows(
  gut_H3K9me3_increment,
  gut_H3K9me3_depletion,
  gut_H3K27ac_increment,
  gut_H3K27ac_depletion,
  gut_Mix,
  gut_Bivalent_increment,
  gut_Bivalent_depletion,
  gut_No_enrichment
)


gut_No_consistent <- data.frame(tissue = "Gut", type = "No consistent", n = 2325-sum(gut$n))

gut <- bind_rows(gut, gut_No_consistent)


ovary_increment_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "increment",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

ovary_depletion_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "depletion",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

ovary_increment_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "increment",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

ovary_depletion_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "depletion",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

ovary_mix_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "mix") %>%
  reframe(unique(TE)) %>% pull()

ovary_increment_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "increment",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

ovary_depletion_bivalent_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "depletion",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

ovary_no_enrichment_TE_list <- epigeneticEffectsTes %>% filter(tissue == "ovary",
                                effect == "noEnrichment",
                                !TE %in% ovary_depletion_H3K27ac_TE_list,
                                !TE %in% ovary_increment_H3K27ac_TE_list,
                                !TE %in% ovary_depletion_H3K9me3_TE_list,
                                !TE %in% ovary_increment_H3K9me3_TE_list,
                                !TE %in% ovary_mix_TE_list) %>%
  reframe(unique(TE)) %>% pull()

upset(fromList(list(ovary_increment_H3K9me3_TE_list=ovary_increment_H3K9me3_TE_list,ovary_depletion_H3K9me3_TE_list=ovary_depletion_H3K9me3_TE_list,ovary_increment_H3K27ac_TE_list=ovary_increment_H3K27ac_TE_list,ovary_depletion_H3K27ac_TE_list=ovary_depletion_H3K27ac_TE_list,ovary_mix_TE_list=ovary_mix_TE_list,ovary_increment_bivalent_TE_list=ovary_increment_bivalent_TE_list,ovary_depletion_bivalent_TE_list=ovary_depletion_bivalent_TE_list,ovary_no_enrichment_TE_list=ovary_no_enrichment_TE_list)), nsets = 20)

# Create data frames for each condition
ovary_H3K9me3_increment <- data.frame(tissue = "Ovary", type = "H3K9me3 increment", n = length(ovary_increment_H3K9me3_TE_list))
ovary_H3K9me3_depletion <- data.frame(tissue = "Ovary", type = "H3K9me3 depletion", n = length(ovary_depletion_H3K9me3_TE_list))
ovary_H3K27ac_increment <- data.frame(tissue = "Ovary", type = "H3K27ac increment", n = length(ovary_increment_H3K27ac_TE_list))
ovary_H3K27ac_depletion <- data.frame(tissue = "Ovary", type = "H3K27ac depletion", n = length(ovary_depletion_H3K27ac_TE_list))
ovary_Mix <- data.frame(tissue = "Ovary", type = "Mix", n = length(ovary_mix_TE_list)-2)
ovary_Bivalent_increment <- data.frame(tissue = "Ovary", type = "Bivalent increment", n = length(ovary_increment_bivalent_TE_list))
ovary_Bivalent_depletion <- data.frame(tissue = "Ovary", type = "Bivalent depletion", n = length(ovary_depletion_bivalent_TE_list))
ovary_No_enrichment <- data.frame(tissue = "Ovary", type = "No enrichment", n = length(ovary_no_enrichment_TE_list))

# Combine the data frames into one
ovary <- bind_rows(
  ovary_H3K9me3_increment,
  ovary_H3K9me3_depletion,
  ovary_H3K27ac_increment,
  ovary_H3K27ac_depletion,
  ovary_Mix,
  ovary_Bivalent_increment,
  ovary_Bivalent_depletion,
  ovary_No_enrichment
)


ovary_No_consistent <- data.frame(tissue = "Ovary", type = "No consistent", n = 2325-sum(ovary$n))

ovary <- bind_rows(ovary, ovary_No_consistent)


type_epigenetic_effects <- rbind(head, gut, ovary)

type_epigenetic_effects$tissue <- factor(type_epigenetic_effects$tissue, levels = c("Head", "Gut", "Ovary"))
type_epigenetic_effects$type <- factor(type_epigenetic_effects$type, levels = rev(c("No enrichment", "H3K9me3 increment", "H3K9me3 depletion", "H3K27ac increment", "H3K27ac depletion", "Bivalent increment", "Bivalent depletion", "Mix", "No consistent")))

ggplot(type_epigenetic_effects, aes(x=tissue,y=n,fill=type)) + geom_col() +
    theme_hc(base_size = 16) +
    labs(x="", y = "Number of TEs", fill = "Type") + theme(
        text = element_text(color = "black"),
        axis.text = element_text(color = "black")) + scale_fill_ptol()  +
    geom_text(aes(label = n),
              position = position_stack(vjust = .5), color="white")


increment_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "increment",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

depletion_H3K9me3_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "depletion",
                                histone == "H3K9me3") %>%
  reframe(unique(TE)) %>% pull()

increment_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "increment",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

depletion_H3K27ac_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "depletion",
                                histone == "H3K27ac") %>%
  reframe(unique(TE)) %>% pull()

mix_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "mix") %>%
  reframe(unique(TE)) %>% pull()

increment_bivalent_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "increment",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

depletion_bivalent_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "depletion",
                                histone == "bivalent") %>%
  reframe(unique(TE)) %>% pull()

no_enrichment_TE_list <- epigeneticEffectsTes %>% filter(
                                effect == "noEnrichment",
                                !TE %in% depletion_H3K27ac_TE_list,
                                !TE %in% increment_H3K27ac_TE_list,
                                !TE %in% depletion_H3K9me3_TE_list,
                                !TE %in% increment_H3K9me3_TE_list,!TE %in% mix_TE_list) %>%
  reframe(unique(TE)) %>% pull()

upset(fromList(list(increment_H3K9me3_TE_list=increment_H3K9me3_TE_list,depletion_H3K9me3_TE_list=depletion_H3K9me3_TE_list,increment_H3K27ac_TE_list=increment_H3K27ac_TE_list,depletion_H3K27ac_TE_list=depletion_H3K27ac_TE_list,mix_TE_list=mix_TE_list,increment_bivalent_TE_list=increment_bivalent_TE_list,depletion_bivalent_TE_list=depletion_bivalent_TE_list,no_enrichment_TE_list=no_enrichment_TE_list)), nsets = 20)

upset(fromList(list(increment_H3K9me3_TE_list=increment_H3K9me3_TE_list,depletion_H3K9me3_TE_list=depletion_H3K9me3_TE_list,increment_H3K27ac_TE_list=increment_H3K27ac_TE_list,depletion_H3K27ac_TE_list=depletion_H3K27ac_TE_list,increment_bivalent_TE_list=increment_bivalent_TE_list,depletion_bivalent_TE_list=depletion_bivalent_TE_list)), nsets = 20, text.scale = 2,nintersects = NA)

```

Metastable TEs
```{r}
dynamic_TEs <- epigeneticEffectsTes %>% filter(effect=="increment",frequency==1 ) %>% dplyr::select(TE,tissue,histone) %>% unique()   %>% group_by(TE) %>%   mutate(tissues = paste0(tissue, collapse = ","),nTissues=(length(unique(tissue))), histones=paste0(unique(sort(histone)),collapse=",")) %>% filter(nTissues>=2) %>% dplyr::select(TE,tissues,histones) %>% unique() %>% group_by(histones)

dynamic_TEs %>% count()

metastable_TEs<-dynamic_TEs[dynamic_TEs$histones=="H3K27ac" | dynamic_TEs$histones=="H3K9me3" | dynamic_TEs$histones=="bivalent",]
metastable_TEs_list <- unique(metastable_TEs$TE)

```

Figure effects
```{r}
library(ggalluvial)
epigeneticEffectsTesAlluvial <- epigeneticEffectsTes[epigeneticEffectsTes$effect=="increment",]
epigeneticEffectsTesAlluvial <- unique(epigeneticEffectsTesAlluvial[,c("TE","tissue","histone")])
count<-epigeneticEffectsTesAlluvial%>% group_by(TE,histone) %>% count()
epigeneticEffectsTesAlluvial <- merge(epigeneticEffectsTesAlluvial,count,by=c("TE","histone"))
epigeneticEffectsTesAlluvial$n<-as.factor(epigeneticEffectsTesAlluvial$n)
epigeneticEffectsTesAlluvial[epigeneticEffectsTesAlluvial$n==1,]$n<-"Body-part specific"
epigeneticEffectsTesAlluvial[epigeneticEffectsTesAlluvial$n==2 | epigeneticEffectsTesAlluvial$n==3 ,]$n<-"Shared"
epigeneticEffectsTesAlluvial$n<-factor(epigeneticEffectsTesAlluvial$n,levels=c("Body-part specific","Shared"))
alluvialIncrement<-arrange(epigeneticEffectsTesAlluvial,desc(n),histone) %>% group_by(histone,tissue,n) %>% count()
alluvialIncrement$tissue<-factor(alluvialIncrement$tissue,levels=c("ovary","gut","head"), labels = c("Ovary","Gut","Head"))
alluvialIncrement$histone<-factor(alluvialIncrement$histone,levels=c("H3K9me3","H3K27ac","bivalent"), labels = c("H3K9me3","H3K27ac","Bivalent"))

alluvialIncrementPlot<-ggplot(alluvialIncrement, aes(y=nn,axis1=histone,axis2=tissue, label = nn)) + 
    geom_alluvium(aes(alpha=n,fill=histone)) + 
    geom_stratum(color="white",fill="gray85") +
    geom_text(stat = "stratum",  aes(label = after_stat(stratum)), size = 4) + 
    theme_minimal(base_size = 18) + scale_alpha_discrete(range = c(0.7, 0.4))   + scale_x_continuous(
        breaks       =  1:2,
        labels       =  c('Histone', 'Body part')) + scale_fill_viridis_d()  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10))  + scale_color_viridis_d()+
    theme(legend.position = "none",   panel.grid.minor = element_blank(),
          axis.text = element_text( face = "bold", color="black", size = 18), axis.text.y=element_blank(), panel.grid.major = element_blank()) + labs(y="")
alluvialIncrementPlot

epigeneticEffectsTesAlluvial <- epigeneticEffectsTes[epigeneticEffectsTes$effect=="depletion",]
epigeneticEffectsTesAlluvial <- unique(epigeneticEffectsTesAlluvial[,c("TE","tissue","histone")])
count<-epigeneticEffectsTesAlluvial%>% group_by(TE,histone) %>% count()
epigeneticEffectsTesAlluvial <- merge(epigeneticEffectsTesAlluvial,count,by=c("TE","histone"))
epigeneticEffectsTesAlluvial$n<-as.factor(epigeneticEffectsTesAlluvial$n)
epigeneticEffectsTesAlluvial[epigeneticEffectsTesAlluvial$n==1,]$n<-"Body-part specific"
epigeneticEffectsTesAlluvial[epigeneticEffectsTesAlluvial$n==2 | epigeneticEffectsTesAlluvial$n==3 ,]$n<-"Shared"
epigeneticEffectsTesAlluvial$n<-factor(epigeneticEffectsTesAlluvial$n,levels=c("Body-part specific","Shared"))
alluvialDepletion<-arrange(epigeneticEffectsTesAlluvial,desc(n),histone) %>% group_by(histone,tissue,n) %>% count()
alluvialDepletion$tissue<-factor(alluvialDepletion$tissue,levels=c("ovary","gut","head"), labels = c("Ovary","Gut","Head"))
alluvialDepletion$histone<-factor(alluvialDepletion$histone,levels=c("H3K9me3","H3K27ac","bivalent"), labels = c("H3K9me3","H3K27ac","Bivalent"))

alluvialDepletionPlot<-ggplot(alluvialDepletion, aes(y=nn,axis1=histone,axis2=tissue, label = nn)) + 
    geom_alluvium(aes(alpha=n,fill=histone)) + 
    geom_stratum(color="white",fill="gray85") +
    geom_text(stat = "stratum",  aes(label = after_stat(stratum)),size=4) + 
    theme_minimal(base_size = 18) + scale_alpha_discrete(range = c(0.7, 0.4))   + scale_x_continuous(
        breaks       =  1:2,
        labels       =  c('Histone', 'Body part')) + scale_fill_viridis_d()  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10))  + scale_color_viridis_d()+
    theme(legend.position = "none",   panel.grid.minor = element_blank(),
          axis.text = element_text( face = "bold", color="black", size = 18), axis.text.y=element_blank(), panel.grid.major = element_blank()) + labs(y="")
alluvialDepletionPlot

 epigeneticEffectsTesPlot<-epigeneticEffectsTes
 epigeneticEffectsTesPlot$effect<-factor(epigeneticEffectsTesPlot$effect, levels=c("increment", "depletion"), labels=c("Enrichment", "Depletion"))
  epigeneticEffectsTesPlot$histone<-factor(epigeneticEffectsTesPlot$histone, levels=c("H3K9me3", "H3K27ac", "bivalent"), labels=c("H3K9me3", "H3K27ac","Bivalent"))
    epigeneticEffectsTesPlot$tissue<-factor(epigeneticEffectsTesPlot$tissue, levels=c("ovary", "gut", "head"), labels=c("Ovary", "Gut","Head"))
  

     a<- ggplot(epigeneticEffectsTesPlot %>% filter(effect=="Enrichment") %>% filter(average <= 500), aes(x=histone, y=average, fill=tissue)) + geom_boxplot(alpha=0.8,position=position_dodge(0.8)) + theme_light(base_size = 18) + scale_fill_tableau(direction=-1) + labs(x="Histone",y="Mean % increase in enrichment", fill="Body part")   + scale_y_continuous(expand = c(0.01,0), breaks = scales::pretty_breaks(n = 10))  +   theme(axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) #+ geom_text(data=countsa, aes(label=paste0("N: ",n), y=-10), position=position_dodge(0.8), hjust=0.5) 
  b<-ggplot(epigeneticEffectsTesPlot %>% filter(effect=="Enrichment"), aes(x=histone, y=spread, fill=tissue)) + geom_boxplot(alpha=0.8,position=position_dodge(0.8)) + theme_light(base_size = 18) + scale_fill_tableau(direction=-1) + labs(x="Histone",y="Mean extent of spread (kb)", fill="Body part")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10))  +   theme(axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) 
  
  c<- ggplot(epigeneticEffectsTesPlot %>% filter(effect=="Depletion"), aes(x=histone, y=average, fill=tissue)) + geom_boxplot(alpha=0.8,position=position_dodge(0.8)) + theme_light(base_size = 18) + scale_fill_tableau(direction=-1) + labs(x="Histone",y="Mean % increase in enrichment", fill="Body part")   + scale_y_continuous(expand = c(0,1), breaks = scales::pretty_breaks(n = 10))  +   theme(axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) #+ geom_text(data=countsc, aes(label=paste0("N: ",n), y=-103), position=position_dodge(0.8), hjust=0.5) 
 c
   d<-ggplot(epigeneticEffectsTesPlot %>% filter(effect=="Depletion"), aes(x=histone, y=spread, fill=tissue)) + geom_boxplot(alpha=0.8,position=position_dodge(0.8)) + theme_light(base_size = 18) + scale_fill_tableau(direction=-1) + labs(x="Histone",y="Mean extent of spread (kb)", fill="Body part")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10))  +   theme(axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) 
d
#plot_grid(b,a,d,c,labels=c("A","B","C","D"),ncol = 2)   
part_a<-ggarrange(alluvialIncrementPlot,alluvialDepletionPlot, ncol=1, nrow=2, labels=c("A", "D"), font.label = list(size = 20))
part_b<-ggarrange(a,b,c,d, ncol=2, nrow=2, common.legend = TRUE, legend="bottom", labels=c("B", "C", "E", "F"), font.label = list(size = 20))
         
part_a<-ggarrange(alluvialIncrementPlot,a+guides(fill="none"),b+guides(fill="none"),align = "hv", widths = c(1,1.4,1.4), nrow=1, labels=c("A", "B", "C"), font.label = list(size = 20))
part_b<-ggarrange(alluvialDepletionPlot,c+guides(fill="none"),d+guides(fill="none"),align = "hv", widths = c(1,1.4,1.4), nrow=1, labels=c("D", "E", "F"), font.label = list(size = 20))
          
ggsave("FIGURES/Figure_2.pdf", plot =  ggarrange(part_a, part_b, align="hv", nrow = 2), width = 16, height = 11)
ggsave("FIGURES/Figure_2_legend.pdf", plot =  a, width = 10, height = 10)

save(file="FIGURES/Figure_2.RData", part_a,part_b)
```

```{r}
epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3",]

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="H3K27ac", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$effect=="depletion" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$effect=="depletion" & epigeneticEffectsTes$histone=="H3K9me3", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$average)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="head" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread)

wilcox.test(unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="ovary" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread, unique(epigeneticEffectsTes[epigeneticEffectsTes$tissue=="gut" & epigeneticEffectsTes$effect=="increment" & epigeneticEffectsTes$histone=="bivalent", c("TE", "average", "spread")])$spread)

# Proportion in gut and ovary that enrich for bivalent
prop.test(c(147,133),n=c(304,304))

# Proportion of H3K9me3 and H3K27ac that deplete
prop.test(c(200,223),n=c(552,552))
# By body part
prop.test(c(267,287),n=c(552,552))
prop.test(c(267,271),n=c(552,552))
prop.test(c(287,271),n=c(552,552))

```

Table 1
```{r}
epigeneticEffectsTes %>% filter(effect=="depletion" ) %>% dplyr::select(histone,TE,spread,average,effect) %>% unique ()%>% group_by(histone, effect) %>% summarize(TEs=length(unique(TE)),meanSpread=mean(spread), sdSpread=sd(spread), meanAverage=mean(average), sdAverage=sd(average)) %>% mutate(spread = paste0(round(meanSpread,2), " (±", round(sdSpread,2),")"), enrichment = paste0(round(meanAverage,2), " (±", round(sdAverage,2),")")) %>% dplyr::select(histone,TEs,spread,enrichment)

epigeneticEffectsTes %>% filter(effect=="increment" ) %>% dplyr::select(histone,TE,spread,average,effect) %>% unique ()%>% group_by(histone, effect) %>% summarize(TEs=length(unique(TE)),meanSpread=mean(spread), sdSpread=sd(spread), meanAverage=mean(average), sdAverage=sd(average)) %>% mutate(spread = paste0(round(meanSpread,2), " (±", round(sdSpread,2),")"), enrichment = paste0(round(meanAverage,2), " (±", round(sdAverage,2),")")) %>% dplyr::select(histone,TEs,spread,enrichment)

epigeneticEffectsTes %>% filter(effect=="depletion" ) %>% dplyr::select(histone,TE,spread,average,effect,tissue) %>% unique ()%>% group_by(histone, effect,tissue) %>% summarize(TEs=length(unique(TE)),meanSpread=mean(spread), sdSpread=sd(spread), meanAverage=mean(average), sdAverage=sd(average)) %>% mutate(spread = paste0(round(meanSpread,2), " (±", round(sdSpread,2),")"), enrichment = paste0(round(meanAverage,2), " (±", round(sdAverage,2),")")) %>% dplyr::select(histone,tissue,TEs,spread,enrichment)

epigeneticEffectsTes %>% filter(effect=="increment" ) %>% dplyr::select(histone,TE,spread,average,effect,tissue) %>% unique ()%>% group_by(histone, effect,tissue) %>% summarize(TEs=length(unique(TE)),meanSpread=mean(spread), sdSpread=sd(spread), meanAverage=mean(average), sdAverage=sd(average)) %>% mutate(spread = paste0(round(meanSpread,2), " (±", round(sdSpread,2),")"), enrichment = paste0(round(meanAverage,2), " (±", round(sdAverage,2),")")) %>% dplyr::select(histone,tissue,TEs,spread,enrichment)

```

```{r dynamic and metastable figure 3}

alluvialBivalent <- fread("epigenetic_effects/alluvialBivalent.txt")

alluvialBivalent$histone1<-factor(alluvialBivalent$histone1,levels=c("H3K9me3","H3K27ac","Bivalent"))
alluvialBivalent$histone2<-factor(alluvialBivalent$histone2,levels=c("H3K9me3","H3K27ac","Bivalent"))

alluvialBivalentPlot<-ggplot(alluvialBivalent, aes(y=nn,axis1=histone1,axis2=histone2, label = nn)) + 
    geom_alluvium(aes(fill=histone1), alpha = 0.7) + 
    geom_stratum(color="white",fill="gray85") +
    geom_text(stat = "stratum",  aes(label = after_stat(stratum)), size = 4) + 
    theme_minimal() +    scale_x_continuous(
        breaks       =  1:2,
        labels       =  c('Histone', 'Histone')) + scale_fill_viridis_d()  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10))  + scale_color_viridis_d()+
    theme(legend.position = "none",   panel.grid.minor = element_blank(),
          axis.text = element_text( face = "bold", color="black", size = 18), axis.text.y=element_blank(), panel.grid.major = element_blank()) + labs(y="")
alluvialBivalentPlot

ggsave("FIGURES/Figure_3.pdf", plot =  alluvialBivalentPlot, width = 6, height = 6)

save(file="FIGURES/Figure_3.RData", alluvialBivalentPlot,alluvialBivalent)
```

Figure S3
```{r figure s3}
      
  freqEnrichmentH3K9me3superfamily <- TEinfo2235 %>% dplyr::select(superfamily,order,class,enrichmentH3K9me3) %>%
    group_by(superfamily) %>%
    summarise(nAll=n(), n = sum(enrichmentH3K9me3==1), order=unique(order),class=unique(class)) %>%
mutate(freq = n / nAll)
  
  ### figure S3
library(tidytext)
 a<-   ggplot(freqEnrichmentH3K9me3superfamily[freqEnrichmentH3K9me3superfamily$superfamily!="INE-1",], aes(x=reorder(superfamily, freq), y=freq, fill=order)) + geom_col(alpha=0.7)+
    geom_text(
        aes(x = superfamily, y = 0, label = paste0(n,"/",nAll)),
        hjust = -0.1,    angle = 90, size = 3
    ) + theme_light() + scale_fill_canva(palette = "Cool and calm")  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5),labels = function(x) paste0(x*100, "%")) + labs(y="Proportion of TEs with epigenetic effects", x = "Superfamily", fill="Order", title="TEs inducing the enrichment of H3K9me3") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    a
    
      freqEnrichmentH3K27acsuperfamily <- TEinfo2235 %>% dplyr::select(superfamily,order,class,enrichmentH3K27ac) %>%
    group_by(superfamily) %>%
    summarise(nAll=n(), n = sum(enrichmentH3K27ac==1), order=unique(order),class=unique(class)) %>%
mutate(freq = n / nAll)
      
          
      freqEnrichmentBivalentsuperfamily <- TEinfo2235 %>% dplyr::select(superfamily,order,class,enrichmentBivalent) %>%
    group_by(superfamily) %>%
    summarise(nAll=n(), n = sum(enrichmentBivalent==1), order=unique(order),class=unique(class)) %>%
mutate(freq = n / nAll)
      
  freqEnrichmentH3K27ac<- TEinfo2235 %>% dplyr::select(superfamily,family,order,class,enrichmentH3K27ac) %>%
    group_by(family) %>%
    summarise(nAll=n(), n = sum(enrichmentH3K27ac==1), order=unique(order),class=unique(class)) %>%
    mutate(freq = n / nAll)

  ggplot(freqEnrichmentH3K27ac, aes(x=reorder(family, freq), y=freq, fill=order)) + geom_col()+
    geom_text(
        aes(x = family, y = 0, label = paste0(n,"/",nAll)),
        hjust = -0.1,    angle = 90, size = 3
    ) + theme_light() + scale_fill_tableau()  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5),,labels = function(x) paste0(x*100, "%")) + labs(y="Proportion of TEs with epigenetic effects", x = "Family", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
  
    ggplot(freqEnrichmentH3K27acsuperfamily, aes(x=reorder(superfamily, freq), y=freq, fill=order)) + geom_col()+
    geom_text(
        aes(x = superfamily, y = 0, label = paste0(n,"/",nAll)),
        hjust = -0.1,    angle = 90, size = 3
    ) + theme_light() + scale_fill_tableau()  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5),labels = function(x) paste0(x*100, "%")) + labs(y="Proportion of TEs with epigenetic effects", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    
     b<-   ggplot(freqEnrichmentH3K27acsuperfamily[freqEnrichmentH3K27acsuperfamily$superfamily!="INE-1",], aes(x=reorder(superfamily, freq), y=freq, fill=order)) + geom_col(alpha=0.7)+
    geom_text(
        aes(x = superfamily, y = 0, label = paste0(n,"/",nAll)),
        hjust = -0.1,    angle = 90, size = 3
    ) + theme_light() + scale_fill_canva(palette = "Cool and calm")  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5),labels = function(x) paste0(x*100, "%")) + labs(y="Proportion of TEs with epigenetic effects", x = "Superfamily", fill="Order", title="TEs inducing the enrichment of H3K27ac") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    b

       c<-   ggplot(freqEnrichmentBivalentsuperfamily[freqEnrichmentBivalentsuperfamily$superfamily!="INE-1",], aes(x=reorder(superfamily, freq), y=freq, fill=order)) + geom_col(alpha=0.7)+
    geom_text(
        aes(x = superfamily, y = 0, label = paste0(n,"/",nAll)),
        hjust = -0.1,    angle = 90, size = 3
    ) + theme_light() + scale_fill_canva(palette = "Cool and calm")  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5),labels = function(x) paste0(x*100, "%")) + labs(y="Proportion of TEs with epigenetic effects", x = "Superfamily", fill="Order", title="TEs inducing bivalent enrichment") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
    c
    
    epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$order=="TRIM",]$order<-"LTR"
    
  epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$tissue=="head",]$tissue<-"Head"
epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$tissue=="gut",]$tissue<-"Gut"
epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$tissue=="ovary",]$tissue<-"Ovary"
epigeneticEffectsTesInfo$tissue<-factor(epigeneticEffectsTesInfo$tissue,levels=c("Head", "Gut","Ovary"))
 
d<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="H3K9me3" & epigeneticEffectsTesInfo$effect=="increment",],aes(x= reorder_within(superfamily, spread, tissue, median),y=spread, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")   + theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5)) + labs(y="Mean extent of H3K9me3 spread (kb)", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
d

 e<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="H3K27ac" & epigeneticEffectsTesInfo$effect=="increment" & epigeneticEffectsTesInfo$superfamily!="INE-1",],aes(x= reorder_within(superfamily, spread, tissue, median),y=spread, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")   +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5)) + labs(y="Mean extent of H3K27ac spread (kb)", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
 e 
 
 f<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="bivalent" & epigeneticEffectsTesInfo$effect=="increment",],aes(x= reorder_within(superfamily, spread, tissue, median),y=spread, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")   +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5)) + labs(y="Mean extent of bivalent spread (kb)", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
 f 
 
 g<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="H3K9me3" & epigeneticEffectsTesInfo$effect=="increment" & epigeneticEffectsTesInfo$average<=505,],aes(x= reorder_within(superfamily, average, tissue, median),y=average, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")  +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 7)) + labs(y="Mean % increase in H3K9me3 enrichment", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 
 g
 
 h<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="H3K27ac" & epigeneticEffectsTesInfo$effect=="increment" & epigeneticEffectsTesInfo$superfamily!="INE-1" & epigeneticEffectsTesInfo$average<500,],aes(x= reorder_within(superfamily, average, tissue, median),y=average, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")   +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5)) + labs(y="Mean % increase in H3K27ac enrichment", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
h

i<-ggplot(epigeneticEffectsTesInfo[epigeneticEffectsTesInfo$histone=="bivalent" & epigeneticEffectsTesInfo$effect=="increment" ,],aes(x= reorder_within(superfamily, average, tissue, median),y=average, fill=order)) + geom_boxplot() + scale_x_reordered()+ facet_wrap(.~tissue,  scales = "free_x") + theme_light() + scale_fill_canva(palette = "Cool and calm")   +  
    theme(panel.grid.minor.x = element_line(colour="grey90", size=0.5, linetype = "dashed"), panel.grid.major.x = element_blank(), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold"))  + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 5)) + labs(y="Mean % increase in bivalent enrichment", x = "Superfamily", fill="Order") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
i
plot_grid(a,b,c,d,e,f,g,h,i,ncol = 3,nrow = 3)

ggsave("FIGURES/Figure_S3.pdf",ggarrange(a,b,c,d,e,f,g,h,i, ncol=3, nrow=3, common.legend = TRUE, legend="bottom", labels=c("A","B","C","D","E","F","G","H","I")),width = 15, height = 13)
save(file="FIGURES/Figure_S3.RData", a,b,c,d,e,f,g,h,i)
```

Expression revised (extending kb)

```{r expression}

TEinfo <- fread("epigenetic_landscapes/TE-clean-breakpoint-frequency.bed")
listY <- read.table("epigenetic_landscapes/listOfProblematicTeY.lst", stringsAsFactors=F)$V1

TEinfo <- TEinfo[! TEinfo$V4 %in% listY, ]
colnames(TEinfo) <- c("chr", "start", "end", "TE", "length","strand","class","order","superfamily","family","strainTE","frequency")
TEinfo<-TEinfo[,c("TE", "length", "class","order","superfamily","family","strainTE")]

epigeneticEffectsTesInfo <- merge(epigeneticEffectsTes,TEinfo,by=c("TE","strainTE"))

epigeneticEffectsTesInfoLength <- merge(epigeneticEffectsTesInfo,unique(TEinfo[,c("TE","length","strainTE")]),by=c("TE","strainTE"))

zscoreTMM <- fread("expression_analysis/zscore_all_pos.tab")
colnames(zscoreTMM) <- c("TE", "gene", "histone", "effect", "tissue", "strainTE", "position", "distance", "zscore", "pvalue", "log2FC")
#zscoreTMM_original <- fread("expression_analysis/zscoreTMM_pos_filtered_original.tab")

  
frequency <- fread("DATA/TE_frequency.tab")


zscoreTMM <- merge(zscoreTMM,frequency[,c("TE","Number_of_genomes_present_(47)", "frequency")])

ggplot(unique(zscoreTMM[,c("TE","frequency")]), aes(x=frequency)) + geom_histogram()
zscoreTMM

epigeneticEffectsEffects <- epigeneticEffectsTesInfo %>% dplyr::select(TE, tissue, strainTE, histone, effect, average, length, superfamily, order,spread) %>% group_by(TE, tissue, strainTE, histone, effect) %>% summarize(avg=mean(average), length=unique(length), superfamily=unique(superfamily), order=unique(order), spread=mean(spread))



Z<-Normal(0,1)
1 - cdf(Z, abs(-37.5)) + cdf(Z, -abs(-37.5))
zscoreTMM[zscoreTMM$pval==0,]$pvalue<-1 - cdf(Z, abs(-37.5)) + cdf(Z, -abs(-37.5))

zscoreTMM<-zscoreTMM[zscoreTMM$log2FC!="Inf" & zscoreTMM$log2FC!="NaN" & !is.na(zscoreTMM$log2FC),]

length(unique(zscoreTMM$TE))
length(unique(zscoreTMM$gene))
length(unique(zscoreTMM[zscoreTMM$zscore>0 & zscoreTMM$pvalue<0.05,]$TE))
length(unique(zscoreTMM[zscoreTMM$zscore<0 & zscoreTMM$pvalue<0.05,]$TE))
table(unique(zscoreTMM[zscoreTMM$pvalue<0.05, c("gene","position")])$position)

ggplot(zscoreTMM, aes(x=log2FC, y=-log10(pvalue))) + geom_point(aes(color = zscore < 0, shape = pvalue < 0.05)) + ylim(c(0,35)) + xlim(-20,20)

zscoreTMM<-merge(zscoreTMM, epigeneticEffectsEffects,by=c("TE","tissue","histone","effect","strainTE"))
zscoreTMM[zscoreTMM$histone=="bivalent",]$histone<-"Bivalent"
zscoreTMM[zscoreTMM$position=="upstream",]$position<-"Upstream"
zscoreTMM[zscoreTMM$position=="downstream",]$position<-"Downstream"
zscoreTMM[zscoreTMM$position=="within",]$position<-"Within gene"
zscoreTMM[zscoreTMM$position=="5UTR",]$position<-"5'UTR"
zscoreTMM[zscoreTMM$position=="3UTR",]$position<-"3'UTR"
zscoreTMM[zscoreTMM$position=="CDS",]$position<-"CDS"
zscoreTMM[zscoreTMM$position=="intron",]$position<-"Intron"

zscoreTMM$histone <- factor(zscoreTMM$histone,levels=c("H3K9me3","H3K27ac","Bivalent"))
zscoreTMM$position <- factor(zscoreTMM$position,levels=c("Upstream", "5'UTR", "CDS", "Intron", "3'UTR", "Downstream"))

expr_plot <- ggplot(zscoreTMM[zscoreTMM$avg<1000,], aes(x=zscore, y=avg)) +

  geom_point(aes(color = position, alpha = pvalue < 0.05)) + 
    theme_light(base_size = 18) + scale_color_manual(values=c("#374E55FF", "#DF8F44FF", "#00A1D5FF", "#79AF97FF", "#B24745FF", "#6A6599FF")) +     scale_alpha_manual(values = c(0.1, 1), guide=FALSE) + 
    labs(x="Expression z-score", y = "Mean % increase in enrichment", color="TE position")   + scale_y_continuous(expand = c(0,0), breaks = scales::pretty_breaks(n = 10)) + theme(axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + facet_grid(.~histone,scales = "free_y") + geom_hline(yintercept=0) + geom_vline(xintercept = 0) + theme(legend.position = "bottom")
expr_plot
ggsave("FIGURES/Figure_5.pdf",expr_plot,width = 10, height = 9)
save(file="FIGURES/Figure_5.RData", expr_plot)

zscoreTMM$pvalueF<-""
zscoreTMM$zscoreF<-""
zscoreTMM[zscoreTMM$pvalue<0.05]$pvalueF <- "Significant"
zscoreTMM[zscoreTMM$pvalue>=0.05]$pvalueF <- "Non-significant"
zscoreTMM[zscoreTMM$zscore>0]$zscoreF <- "Positive"
zscoreTMM[zscoreTMM$zscore<0]$zscoreF <- "Negative"

length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscore < 0,]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscore > 0,]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscore < 0 & zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscore > 0 & zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="Bivalent"]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment"& zscoreTMM$zscore < 0,]$TE))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant"& zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="depletion",]$TE))

length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$tissue=="head",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$tissue=="gut",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$tissue=="ovary",]$gene))
length(unique(zscoreTMM[ zscoreTMM$tissue=="head",]$gene))
length(unique(zscoreTMM[ zscoreTMM$tissue=="gut",]$gene))
length(unique(zscoreTMM[ zscoreTMM$tissue=="ovary",]$gene))
# head
228/482
# gut
305/746
# ovary
427/879

prop.test(x=c(228,305), n=c(482,746))
prop.test(x=c(228,427), n=c(482,879))
prop.test(x=c(305,427), n=c(746,879))

length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$strainTE=="AKA-017",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$strainTE=="JUT-011",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$strainTE=="MUN-016",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$strainTE=="SLA-001",]$gene))
length(unique(zscoreTMM[ zscoreTMM$pvalueF=="Significant" & zscoreTMM$strainTE=="TOM-007",]$gene))
prop.test(x=c(228,305), n=c(482,746))

write.table(zscoreTMM,file="FIGURES/zscoreTMM.tab", col.names = T, sep = "\t", row.names = F, quote = F)

```


```{r table2}
length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="depletion",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="depletion",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="depletion",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Positive",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="increment" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))

length(unique(zscoreTMM[zscoreTMM$histone=="H3K9me3" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="H3K27ac" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))
length(unique(zscoreTMM[zscoreTMM$histone=="Bivalent" & zscoreTMM$effect=="depletion" & zscoreTMM$pvalueF=="Significant" & zscoreTMM$zscoreF=="Negative",]$TE))
```

Overall, we annotated 28,947 TEs across the 47 genomes

```{r freq}

frequency <- fread("DATA/TE_frequency.tab")
dist <- fread("DATA/TE_gene_distance.tab")
frequency <- merge(frequency, dist)
frequency <- as.data.frame(frequency)
epigeneticEffectsTes2 <- merge(epigeneticEffectsTes, frequency[,c("TE","GeneDistance","Number_of_genomes_present_(47)")])

TEs_enrich_H3K9me3 <- epigeneticEffectsTes2[epigeneticEffectsTes2$effect=="increment" & epigeneticEffectsTes2$histone=="H3K9me3",] # 816 TEs
TEs_enrich_H3K9me3 <- unique(TEs_enrich_H3K9me3[,c("TE","GeneDistance","Number_of_genomes_present_(47)")])
summary(TEs_enrich_H3K9me3$`Number_of_genomes_present_(47)`)
summary(frequency$`Number_of_genomes_present_(47)`)

TEs_enrich_H3K27ac <- epigeneticEffectsTes2[epigeneticEffectsTes2$effect=="increment" & epigeneticEffectsTes2$histone=="H3K27ac",] # 345 TEs
TEs_enrich_H3K27ac <- unique(TEs_enrich_H3K27ac[,c("TE","GeneDistance","Number_of_genomes_present_(47)")])

list_TEs_5G <- fread("expression_analysis/TEannotations/TEs_5G_4874.list", header = F)
frequency_5G <- frequency[frequency$TE %in% list_TEs_5G$V1,]

length(unique(frequency_5G[frequency_5G$`Number_of_genomes_present_(47)`==1,]$TE))
length(unique(TEs_enrich_H3K9me3[TEs_enrich_H3K9me3$`Number_of_genomes_present_(47)`==1,]$TE))
fisher.test(matrix(c(4823,2316,816,628),nrow=2))
fisher.test(matrix(c(4823,2946,816,766),nrow=2))
fisher.test(matrix(c(4823,1692,816,30),nrow=2))

length(unique(TEs_enrich_H3K27ac[TEs_enrich_H3K27ac$`Number_of_genomes_present_(47)`==1,]$TE))
length(unique(TEs_enrich_H3K27ac[TEs_enrich_H3K27ac$`Number_of_genomes_present_(47)`<=3,]$TE))
length(unique(TEs_enrich_H3K27ac[TEs_enrich_H3K27ac$`Number_of_genomes_present_(47)`<=5,]$TE))
fisher.test(matrix(c(4823,2316,345,275),nrow=2))
fisher.test(matrix(c(4823,2946,345,325),nrow=2))
fisher.test(matrix(c(4823,3131,345,336),nrow=2))


length(unique(frequency_5G[!is.na(frequency_5G$GeneDistance),]$TE))
fisher.test(matrix(c(3528,1295,630,186),nrow=2))


zscoreTMM2 <- merge(zscoreTMM, dist)
TEs_enrich_H3K9me3_downregulate <- zscoreTMM2[zscoreTMM2$zscoreF=="Negative" & zscoreTMM2$pvalueF=="Significant" & zscoreTMM2$histone=="H3K9me3" & zscoreTMM2$effect=="increment",]

TEs_enrich_H3K9me3_downregulate <- zscoreTMM2[zscoreTMM2$zscoreF=="Negative" & zscoreTMM2$pvalueF=="Significant" & zscoreTMM2$histone=="H3K9me3" & zscoreTMM2$effect=="increment",]

length(unique(TEs_enrich_H3K9me3_downregulate$TE))
length(unique(TEs_enrich_H3K9me3_downregulate[TEs_enrich_H3K9me3_downregulate$`Number_of_genomes_present_(47)`==1,]$TE))
length(unique(TEs_enrich_H3K9me3_downregulate[TEs_enrich_H3K9me3_downregulate$`Number_of_genomes_present_(47)`<=3,]$TE))
length(unique(TEs_enrich_H3K9me3_downregulate[TEs_enrich_H3K9me3_downregulate$`Number_of_genomes_present_(47)`<=5,]$TE))

fisher.test(matrix(c(4823,2316,238,173),nrow=2))
fisher.test(matrix(c(4823,2946,238,220),nrow=2))
fisher.test(matrix(c(4823,3131,238,229),nrow=2))

TEs_enrich_H3K27ac_upregulate <- zscoreTMM2[zscoreTMM2$zscoreF=="Positive" & zscoreTMM2$pvalueF=="Significant" & zscoreTMM2$histone=="H3K27ac" & zscoreTMM2$effect=="increment",]
length(unique(TEs_enrich_H3K27ac_upregulate[TEs_enrich_H3K27ac_upregulate$`Number_of_genomes_present_(47)`==1,]$TE))
length(unique(TEs_enrich_H3K27ac_upregulate[TEs_enrich_H3K27ac_upregulate$`Number_of_genomes_present_(47)`<=3,]$TE))
length(unique(TEs_enrich_H3K27ac_upregulate[TEs_enrich_H3K27ac_upregulate$`Number_of_genomes_present_(47)`<=5,]$TE))
fisher.test(matrix(c(4823,2316,51,44),nrow=2))
fisher.test(matrix(c(4823,2946,51,48),nrow=2))
fisher.test(matrix(c(4823,3131,51,49),nrow=2))

TEs_enrich_H3K9me3_effect <- zscoreTMM2[zscoreTMM2$pvalueF=="Significant" & zscoreTMM2$histone=="H3K9me3" & zscoreTMM2$effect=="increment",]
length(unique(TEs_enrich_H3K9me3_effect[TEs_enrich_H3K9me3_effect$`Number_of_genomes_present_(47)`==1,]$TE))
fisher.test(matrix(c(4823,2316,340,252),nrow=2))
fisher.test(matrix(c(4823,2946,340,317),nrow=2))
fisher.test(matrix(c(4823,3131,340,329),nrow=2))


TEs_enrich_H3K27ac_effect <- zscoreTMM2[ zscoreTMM2$pvalueF=="Significant" & zscoreTMM2$histone=="H3K27ac" & zscoreTMM2$effect=="increment",]
length(unique(TEs_enrich_H3K27ac_effect[TEs_enrich_H3K27ac_effect$`Number_of_genomes_present_(47)`==1,]$TE))
length(unique(TEs_enrich_H3K27ac_effect[TEs_enrich_H3K27ac_effect$`Number_of_genomes_present_(47)`<=3,]$TE))
length(unique(TEs_enrich_H3K27ac_effect[TEs_enrich_H3K27ac_effect$`Number_of_genomes_present_(47)`<=5,]$TE))
fisher.test(matrix(c(4823,2316,102,83),nrow=2))
fisher.test(matrix(c(4823,2946,102,97),nrow=2))
fisher.test(matrix(c(4823,3131,102,98),nrow=2))

```

```{r tes-properties}
zscoreTMM_significant_genes <- zscoreTMM[zscoreTMM$pvalueF=="Significant",]
zscoreTMM_non_significant_genes <- zscoreTMM[zscoreTMM$pvalueF=="Non-significant",]

zscoreTMM_significant_genes_H3K9me3_increment <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="increment" & zscoreTMM_significant_genes$histone=="H3K9me3",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_H3K9me3_increment <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="increment" & zscoreTMM_non_significant_genes$histone=="H3K9me3",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_increment$avg,
            zscoreTMM_significant_genes_H3K9me3_increment$avg)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_increment$length,
            zscoreTMM_significant_genes_H3K9me3_increment$length)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_increment$frequency,
            zscoreTMM_significant_genes_H3K9me3_increment$frequency)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_increment$spread,
            zscoreTMM_significant_genes_H3K9me3_increment$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="increment" & histone=="H3K9me3" & pvalueF=="Significant" | effect=="increment" & histone=="H3K9me3" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,position,gene, pvalueF) %>% unique() 
 
 kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")
  
 zscoreTMM_significant_genes_H3K27ac_increment <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="increment" & zscoreTMM_significant_genes$histone=="H3K27ac",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_H3K27ac_increment <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="increment" & zscoreTMM_non_significant_genes$histone=="H3K27ac",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_increment$avg,
            zscoreTMM_significant_genes_H3K27ac_increment$avg)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_increment$length,
            zscoreTMM_significant_genes_H3K27ac_increment$length)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_increment$frequency,
            zscoreTMM_significant_genes_H3K27ac_increment$frequency)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_increment$spread,
            zscoreTMM_significant_genes_H3K27ac_increment$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="increment" & histone=="H3K27ac" & pvalueF=="Significant" | effect=="increment" & histone=="H3K27ac" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,gene,position, pvalueF) %>% unique() 
 
  kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")

zscoreTMM_significant_genes_bivalent_increment <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="increment" & zscoreTMM_significant_genes$histone=="Bivalent",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_bivalent_increment <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="increment" & zscoreTMM_non_significant_genes$histone=="Bivalent",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_bivalent_increment$avg,
            zscoreTMM_significant_genes_bivalent_increment$avg)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_increment$length,
            zscoreTMM_significant_genes_bivalent_increment$length)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_increment$frequency,
            zscoreTMM_significant_genes_bivalent_increment$frequency)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_increment$spread,
            zscoreTMM_significant_genes_bivalent_increment$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="increment" & histone=="Bivalent" & pvalueF=="Significant" | effect=="increment" & histone=="Bivalent" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,position, pvalueF) %>% unique() 
 
  kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")

   zscoreTMM_significant_genes_H3K9me3_depletion <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="depletion" & zscoreTMM_significant_genes$histone=="H3K9me3",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_H3K9me3_depletion <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="depletion" & zscoreTMM_non_significant_genes$histone=="H3K9me3",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_depletion$avg,
            zscoreTMM_significant_genes_H3K9me3_depletion$avg)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_depletion$length,
            zscoreTMM_significant_genes_H3K9me3_depletion$length)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_depletion$frequency,
            zscoreTMM_significant_genes_H3K9me3_depletion$frequency)
wilcox.test(zscoreTMM_non_significant_genes_H3K9me3_depletion$spread,
            zscoreTMM_significant_genes_H3K9me3_depletion$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="depletion" & histone=="H3K9me3" & pvalueF=="Significant" | effect=="depletion" & histone=="H3K9me3" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,position,gene, pvalueF) %>% unique() 
  kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")

 zscoreTMM_significant_genes_H3K27ac_depletion <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="depletion" & zscoreTMM_significant_genes$histone=="H3K27ac",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_H3K27ac_depletion <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="depletion" & zscoreTMM_non_significant_genes$histone=="H3K27ac",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_depletion$avg,
            zscoreTMM_significant_genes_H3K27ac_depletion$avg)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_depletion$length,
            zscoreTMM_significant_genes_H3K27ac_depletion$length)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_depletion$frequency,
            zscoreTMM_significant_genes_H3K27ac_depletion$frequency)
wilcox.test(zscoreTMM_non_significant_genes_H3K27ac_depletion$spread,
            zscoreTMM_significant_genes_H3K27ac_depletion$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="depletion" & histone=="H3K27ac" & pvalueF=="Significant" | effect=="depletion" & histone=="H3K27ac" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,gene,position, pvalueF) %>% unique() 
 
  kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")

zscoreTMM_significant_genes_bivalent_depletion <- unique(zscoreTMM_significant_genes[zscoreTMM_significant_genes$effect=="depletion" & zscoreTMM_significant_genes$histone=="Bivalent",c("TE", "avg", "frequency", "length", "spread")])
zscoreTMM_non_significant_genes_bivalent_depletion <- unique(zscoreTMM_non_significant_genes[zscoreTMM_non_significant_genes$effect=="depletion" & zscoreTMM_non_significant_genes$histone=="Bivalent",c("TE", "avg", "frequency", "length", "spread")])
wilcox.test(zscoreTMM_non_significant_genes_bivalent_depletion$avg,
            zscoreTMM_significant_genes_bivalent_depletion$avg)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_depletion$length,
            zscoreTMM_significant_genes_bivalent_depletion$length)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_depletion$frequency,
            zscoreTMM_significant_genes_bivalent_depletion$frequency)
wilcox.test(zscoreTMM_non_significant_genes_bivalent_depletion$spread,
            zscoreTMM_significant_genes_bivalent_depletion$spread)
  
 
 kk<-zscoreTMM%>%filter(effect=="depletion" & histone=="Bivalent" & pvalueF=="Significant" | effect=="depletion" & histone=="Bivalent" & pvalueF=="Non-significant" )  %>% dplyr::select(TE,position, pvalueF) %>% unique() 
 
  kk$position_grouped <- ""
 kk[kk$position=="5'UTR" | kk$position=="CDS" | kk$position=="3'UTR",]$position_grouped <- "Exon"
 kk[kk$position=="Upstream",]$position_grouped <- "Upstream"
    kk[kk$position=="Downstream",]$position_grouped <- "Downstream"
    kk[kk$position=="Intron",]$position_grouped <- "Intron"

 chisq.test(kk$position_grouped, kk$pvalueF, correct=FALSE)
  table(kk$position_grouped, kk$pvalueF)
chisq.posthoc.test(as.matrix(table(kk$pvalueF, kk$position_grouped )), simulate.p.value = TRUE, method = "bonferroni")
```

Enrichment families

```{r enrichment-families}

 # info 2325
 
TEinfo <- fread("epigenetic_landscapes/TE-clean-breakpoint-frequency.bed")
TEinfo <- TEinfo[! TEinfo$V4 %in% listY, ]
colnames(TEinfo) <- c("chr", "start", "end", "TE", "length","strand","class","order","superfamily","family","strainTE","frequency")
TEinfo2325<-unique(TEinfo[,c("TE",  "class","order","superfamily","family","frequency")])

TEs_exclude <- unique(epigeneticEffectsTes[!epigeneticEffectsTes$TE %in% TEs_epigenetic_effects & epigeneticEffectsTes$effect=="mix",]$TE)

TEs_keep <- unique(epigeneticEffectsEffects[!epigeneticEffectsEffects$TE%in%TEs_exclude,]$TE)

TEinfo2235 <- TEinfo2325[TEinfo2325$TE%in%TEs_keep,]


tmp<-epigeneticEffectsTes %>% filter(histone=="H3K9me3", effect=="increment") %>% distinct(TE) %>% pull()
TEinfo2235$enrichmentH3K9me3 <- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
tmp<-epigeneticEffectsTes %>% filter(histone=="H3K27ac", effect=="increment") %>% distinct(TE) %>% pull()
TEinfo2235$enrichmentH3K27ac <- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
tmp<-epigeneticEffectsTes %>% filter(histone=="H3K9me3", effect=="depletion") %>% distinct(TE) %>% pull()
TEinfo2235$depletionH3K9me3 <- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
tmp<-epigeneticEffectsTes %>% filter(histone=="H3K27ac", effect=="depletion") %>% distinct(TE) %>% pull()
TEinfo2235$depletionH3K27ac<- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
tmp<-epigeneticEffectsTes %>% filter(histone=="bivalent", effect=="increment") %>% distinct(TE) %>% pull()
TEinfo2235$enrichmentBivalent<- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
tmp<-epigeneticEffectsTes %>% filter(histone=="bivalent", effect=="depletion") %>% distinct(TE) %>% pull()
TEinfo2235$depletionBivalent<- ifelse(TEinfo2235$TE %in% tmp, '1', '0')
### CUIDADO CON ESTO ###
TEinfo2235[TEinfo2235$order=="TRIM",]$order<-"LTR"


####
ggplot(TEinfo2235, aes(x=superfamily,fill=depletionH3K27ac)) + geom_bar(position="fill") + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))  +
    geom_text(stat='count', aes(label=..count..), vjust=-1, position="fill")



  freqEnrichmentH3K9me3 <- TEinfo2235 %>% dplyr::select(superfamily,family,order,class,enrichmentH3K9me3) %>%
    group_by(family) %>%
    summarise(nAll=n(), n = sum(enrichmentH3K9me3==1), order=unique(order),class=unique(class)) %>%
    mutate(freq = n / nAll)
  
  freqEnrichmentH3K9me3$p.value  <- 0
  freqEnrichmentH3K9me3$sign <- ""
  
  freqEnrichmentH3K9me3_filter <- freqEnrichmentH3K9me3[freqEnrichmentH3K9me3$nAll>=20 & freqEnrichmentH3K9me3$n>7,]
  for ( TE in 1:nrow(freqEnrichmentH3K9me3_filter )) {
    freqEnrichmentH3K9me3_filter[TE,2]
    freqEnrichmentH3K9me3_filter[TE,3]
    p_value <- prop.test(x = c(pull(freqEnrichmentH3K9me3_filter[TE,3]), 816), n = c(pull(freqEnrichmentH3K9me3_filter[TE,2]), 2235))$p.value
    sign <- ifelse(prop.test(x = c(pull(freqEnrichmentH3K9me3_filter[TE,3]), 816), n = c(pull(freqEnrichmentH3K9me3_filter[TE,2]), 2235))$estimate[1] > 816/2235, "+", "-")
    freqEnrichmentH3K9me3_filter[TE,]$p.value<- p_value
    freqEnrichmentH3K9me3_filter[TE,]$sign<- sign
  }
  
  
    freqEnrichmentH3K27ac <- TEinfo2235 %>% dplyr::select(superfamily,family,order,class,enrichmentH3K27ac) %>%
    group_by(family) %>%
    summarise(nAll=n(), n = sum(enrichmentH3K27ac==1), order=unique(order),class=unique(class)) %>%
    mutate(freq = n / nAll)
  
  freqEnrichmentH3K27ac$p.value  <- 0
  freqEnrichmentH3K27ac$sign <- ""
  
  freqEnrichmentH3K27ac_filter <- freqEnrichmentH3K27ac[freqEnrichmentH3K27ac$nAll>=20 & freqEnrichmentH3K27ac$n>7,]
  for ( TE in 1:nrow(freqEnrichmentH3K27ac_filter )) {
    freqEnrichmentH3K27ac_filter[TE,2]
    freqEnrichmentH3K27ac_filter[TE,3]
    p_value <- prop.test(x = c(pull(freqEnrichmentH3K27ac_filter[TE,3]), 345), n = c(pull(freqEnrichmentH3K27ac_filter[TE,2]), 2235))$p.value
    sign <- ifelse(prop.test(x = c(pull(freqEnrichmentH3K27ac_filter[TE,3]), 345), n = c(pull(freqEnrichmentH3K27ac_filter[TE,2]), 2235))$estimate[1] > 345/2235, "+", "-")
    freqEnrichmentH3K27ac_filter[TE,]$p.value<- p_value
    freqEnrichmentH3K27ac_filter[TE,]$sign<- sign
  }
  
    
    freqEnrichmentBivalent <- TEinfo2235 %>% dplyr::select(superfamily,family,order,class,enrichmentBivalent) %>%
    group_by(family) %>%
    summarise(nAll=n(), n = sum(enrichmentBivalent==1), order=unique(order),class=unique(class)) %>%
    mutate(freq = n / nAll)
  
  freqEnrichmentBivalent$p.value  <- 0
  freqEnrichmentBivalent$sign <- ""
  
  freqEnrichmentBivalent_filter <- freqEnrichmentBivalent[freqEnrichmentBivalent$nAll>=20 & freqEnrichmentBivalent$n>7,]
  for ( TE in 1:nrow(freqEnrichmentBivalent_filter )) {
    freqEnrichmentBivalent_filter[TE,2]
    freqEnrichmentBivalent_filter[TE,3]
    p_value <- prop.test(x = c(pull(freqEnrichmentBivalent_filter[TE,3]), 304), n = c(pull(freqEnrichmentBivalent_filter[TE,2]), 2235))$p.value
    sign <- ifelse(prop.test(x = c(pull(freqEnrichmentBivalent_filter[TE,3]), 304), n = c(pull(freqEnrichmentBivalent_filter[TE,2]), 2235))$estimate[1] > 304/2235, "+", "-")
    freqEnrichmentBivalent_filter[TE,]$p.value<- p_value
    freqEnrichmentBivalent_filter[TE,]$sign<- sign
  }
  
  freqEnrichmentH3K9me3_filter$histone <- "H3K9me3"
  freqEnrichmentH3K9me3_filter$fill <- "ns"
  freqEnrichmentH3K9me3_filter[freqEnrichmentH3K9me3_filter$sign=="+" & freqEnrichmentH3K9me3_filter$p.value < 0.05,]$fill <- "sign_pos"
  freqEnrichmentH3K9me3_filter[freqEnrichmentH3K9me3_filter$sign=="-" & freqEnrichmentH3K9me3_filter$p.value < 0.05,]$fill <- "sign_neg"
  freqEnrichmentH3K9me3_filter$fill <- factor(freqEnrichmentH3K9me3_filter$fill, level=c("sign_pos","ns","sign_neg"))
  lvls <- freqEnrichmentH3K9me3_filter$family[order(freqEnrichmentH3K9me3_filter$fill,-freqEnrichmentH3K9me3_filter$freq)]
  freqEnrichmentH3K9me3_filter$family <- factor(freqEnrichmentH3K9me3_filter$family, levels=lvls)


a<-ggplot(freqEnrichmentH3K9me3_filter, aes(x=family, y = freq, fill=fill)) + geom_col() + coord_flip() +     scale_x_discrete(limits=rev) + theme_light() +  
    theme(panel.grid.minor.x = element_blank(), panel.grid.major.x = element_line(colour="grey90", linewidth = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + scale_y_continuous(labels = function(x) paste0(x*100, "%"), expand = c(0,0)) + labs(x="Family", y= "Frequency", fill="Type") + scale_fill_manual(values=c("#D81A60","gray80","#1D89E4")) + geom_hline(yintercept = 0.35) + geom_text(aes(y=0.01, label = order), hjust=0, size=2.5) + guides(fill="none") + annotate(geom = "text", x = 1, y =0.35 ,label="Mean H3K9me3 enrichment",  angle = 90, size=3, hjust = 0, vjust=-1.5)  

freqEnrichmentH3K27ac_filter$histone <- "H3K27ac"
  freqEnrichmentH3K27ac_filter$fill <- "ns"
  freqEnrichmentH3K27ac_filter[freqEnrichmentH3K27ac_filter$sign=="+" & freqEnrichmentH3K27ac_filter$p.value < 0.05,]$fill <- "sign_pos"
  freqEnrichmentH3K27ac_filter[freqEnrichmentH3K27ac_filter$sign=="-" & freqEnrichmentH3K27ac_filter$p.value < 0.05,]$fill <- "sign_neg"
  freqEnrichmentH3K27ac_filter$fill <- factor(freqEnrichmentH3K27ac_filter$fill, level=c("sign_pos","ns","sign_neg"))
  lvls <- freqEnrichmentH3K27ac_filter$family[order(freqEnrichmentH3K27ac_filter$fill,-freqEnrichmentH3K27ac_filter$freq)]
  freqEnrichmentH3K27ac_filter$family <- factor(freqEnrichmentH3K27ac_filter$family, levels=lvls)
  
b<-ggplot(freqEnrichmentH3K27ac_filter, aes(x=family, y = freq, fill=fill)) + geom_col() + coord_flip() +     scale_x_discrete(limits=rev) + theme_light() +  
    theme(panel.grid.minor.x = element_blank(), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + scale_y_continuous(labels = function(x) paste0(x*100, "%"), expand = c(0,0)) + labs(x="", y= "Frequency", fill="Type") + scale_fill_manual(values=c("#D81A60","gray80","#1D89E4")) + geom_hline(yintercept = 0.148) + geom_text(aes(y=0.01, label = order), hjust=0, size=2.5) + guides(fill="none") + annotate(geom = "text", x = 1, y =0.148 ,label="Mean H3K27ac enrichment",  angle = 90, size=3, hjust = 0, vjust=-1.5)  



freqEnrichmentBivalent_filter$histone <- "Bivalent"
  freqEnrichmentBivalent_filter$fill <- "ns"
  freqEnrichmentBivalent_filter[freqEnrichmentBivalent_filter$sign=="+" & freqEnrichmentBivalent_filter$p.value < 0.05,]$fill <- "sign_pos"
  freqEnrichmentBivalent_filter[freqEnrichmentBivalent_filter$sign=="-" & freqEnrichmentBivalent_filter$p.value < 0.05,]$fill <- "sign_neg"
  freqEnrichmentBivalent_filter$fill <- factor(freqEnrichmentBivalent_filter$fill, level=c("sign_pos","ns","sign_neg"))
  lvls <- freqEnrichmentBivalent_filter$family[order(freqEnrichmentBivalent_filter$fill,-freqEnrichmentBivalent_filter$freq)]
  freqEnrichmentBivalent_filter$family <- factor(freqEnrichmentBivalent_filter$family, levels=lvls)
  
c<-ggplot(freqEnrichmentBivalent_filter, aes(x=family, y = freq, fill=fill)) + geom_col() + coord_flip() +     scale_x_discrete(limits=rev) + theme_light() +  
    theme(panel.grid.minor.x = element_blank(), panel.grid.major.x = element_line(colour="grey90", size = 0.5,linetype = "dashed"), panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(), axis.text = element_text(colour = "black"), axis.title = element_text(face = "bold")) + scale_y_continuous(labels = function(x) paste0(x*100, "%"), expand = c(0,0)) + labs(x="", y= "Frequency", fill="Type") + scale_fill_manual(values=c("#D81A60","gray80","#1D89E4")) + geom_hline(yintercept = 0.13) + geom_text(aes(y=0.01, label = order), hjust=0, size=2.5) + guides(fill="none") + annotate(geom = "text", x = 1, y = 0.13 ,label="Mean bivalent enrichment",  angle = 90, size=3, hjust = 0, vjust=-1.5)  

ggarrange(a,b,c, ncol=3, nrow=1, common.legend = TRUE, legend="bottom", labels=c("A","B","C"))

ggsave("FIGURES/Figure_4.pdf", plot = ggarrange(a+ggtitle("H3K9me3 enrichment"),b+ggtitle("H3K27ac enrichment"),c+ggtitle("Bivalent enrichment"), ncol=3, nrow=1, common.legend = TRUE, legend="bottom", labels=c("A","B","C")), width = 9, height = 6)
save(file="FIGURES/Figure_4.RData", a, b, c, freqEnrichmentBivalent_filter, freqEnrichmentH3K27ac_filter, freqEnrichmentH3K9me3_filter)

write.table(rbind(freqEnrichmentH3K9me3_filter,freqEnrichmentH3K27ac_filter,freqEnrichmentBivalent_filter),file = "FIGURES/Figure_4_data.tsv", col.names = T, row.names = F, quote = F, sep = "\t")

```

```{r GO}
tissues <- c("head", "gut", "ovary")
effects <- c("All", "Positive", "Negative")
for (t in tissues) {
for (z in effects) {
if ( z == "All") {
list_genes <- unique(zscoreTMM[zscoreTMM$tissue==t & zscoreTMM$pvalueF=="Significant",]$gene)
} else {
list_genes <- unique(zscoreTMM[zscoreTMM$tissue==t & zscoreTMM$zscoreF == z & zscoreTMM$pvalueF=="Significant",]$gene)
}
write.table(x = as.data.frame(list_genes), file = paste0("GO_analysis/",t,".",z,".TE_genes.lst"), row.names = F, col.names = F, quote = F)

assign(paste0("GO_",t,".",z,".TE_genes.lst"),list_genes)


}
}

write.table(x = as.data.frame(unique(zscoreTMM[zscoreTMM$TE%in%metastable_TEs_list,]$gene)), file = paste0("GO_analysis/metastable.TE_genes.lst"), row.names = F, col.names = F, quote = F)

library(clusterProfiler)
library(org.Dm.eg.db)

head <- fread("GO_analysis/head.all_genes.lst", header = F)$V1
gut <- fread("GO_analysis/gut.all_genes.lst", header = F)$V1
ovary <- fread("GO_analysis/ovary.all_genes.lst", header = F)$V1
all <- fread("GO_analysis/all_genes.lst", header = F)$V1

go_enrich <- clusterProfiler::enrichGO(gene = GO_head.All.TE_genes.lst,
                      universe = head,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
head_all <- dotplot(go_enrich, title = "Head - All genes")
write.table(go_enrich@result, "FIGURES/GO_head_all.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_head.Positive.TE_genes.lst,
                      universe = head,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Head - Upregulated")
head_upregulated <- dotplot(go_enrich, title = "Head - Upregulated")

write.table(go_enrich@result, "FIGURES/GO_head_upregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_head.Negative.TE_genes.lst,
                      universe = head,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Head - Downregulated")
head_downregulated <- dotplot(go_enrich, title = "Head - Downregulated")

write.table(go_enrich@result, "FIGURES/GO_head_downregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_gut.All.TE_genes.lst,
                      universe = gut,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Gut - all genes")
gut_all <- dotplot(go_enrich, title = "Gut - All genes")

write.table(go_enrich@result, "FIGURES/GO_gut_all.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_gut.Positive.TE_genes.lst,
                      universe = gut,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Gut - Upregulated")
gut_upregulated <-dotplot(go_enrich, title = "Gut - Upregulated")

write.table(go_enrich@result, "FIGURES/GO_gut_upregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_gut.Negative.TE_genes.lst,
                      universe = gut,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Gut - Downregulated")
gut_downregulated <-dotplot(go_enrich, title = "Gut - Downregulated")

write.table(go_enrich@result, "FIGURES/GO_gut_downregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_ovary.All.TE_genes.lst,
                      universe = ovary,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Ovary - all genes")
ovary_all <-dotplot(go_enrich, title = "Ovary - All genes")

write.table(go_enrich@result, "FIGURES/GO_ovary_all.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_ovary.Positive.TE_genes.lst,
                      universe = ovary,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Ovary - Upregulated")
ovary_upregulated <-dotplot(go_enrich, title = "Ovary - Upregulated")

write.table(go_enrich@result, "FIGURES/GO_ovary_upregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = GO_ovary.Negative.TE_genes.lst,
                      universe = ovary,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Ovary - Downregulated")
ovary_downregulated <-dotplot(go_enrich, title = "Ovary - Downregulated")

write.table(go_enrich@result, "FIGURES/GO_ovary_downregulated.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

go_enrich <- clusterProfiler::enrichGO(gene = unique(zscoreTMM[zscoreTMM$TE%in%metastable_TEs_list,]$gene),
                      universe = all,
                      OrgDb = org.Dm.eg.db, 
                      keyType = 'FLYBASE',
                      readable = T,
                      ont = "BP",
                      pvalueCutoff = 0.05, 
                      qvalueCutoff = 0.05)
dotplot(go_enrich, title = "Metastable TEs")
metastable <-dotplot(go_enrich, title = "Metastable TEs")

write.table(go_enrich@result, "FIGURES/GO_metastable.tsv", col.names = T, sep = "\t", quote = F, row.names = F)

ggarrange(head_all, head_upregulated, head_downregulated, gut_all, gut_upregulated, gut_downregulated,metastable,labels="AUTO")
ggsave("FIGURES/Figure_S4.pdf", ggarrange(head_all, head_upregulated, head_downregulated, gut_all, gut_upregulated, gut_downregulated,metastable,labels="AUTO"), width = 18, height = 18)
```